lib/internal/async_hooks.js

```diff
@@ -243,13 +244,90 @@ function restoreActiveHooks() {
   active_hooks.tmp_fields = null;
 }
 
+const promiseWrap = Symbol('promiseWrap');
+const destroyedSymbol = Symbol('destroyed');
+
+class PromiseWrap {
+  constructor(promise, parent_wrap, silent) {
+    promise[promiseWrap] = this;
+    this.isChainedPromise = !!parent_wrap;
+    this.promise = promise;
+    this[destroyedSymbol] = undefined;
+
+    const asyncId = getOrSetAsyncId(this);
+
+    if (typeof parent_wrap !== 'undefined') {
+      this.triggerAsyncId = getOrSetAsyncId(parent_wrap);
+    } else {
+      this.triggerAsyncId = getDefaultTriggerAsyncId();
+    }
+
+    if (!silent && initHooksExist()) {
+      emitInitScript(asyncId, 'PROMISE', this.triggerAsyncId, this);
+
+      if (destroyHooksExist()) {
```

Comment:
In scenario when there are no `init` hooks, but at least one active `destroy` hook exists, we still need to emit `destroy` event. So, we should move this block one level up, like it's done in `AsyncResource`. WDYT?

Reply:
Yep, good call. Though I struggle to imagine a scenario when you would not want the init. 🤔

Reply:
Yeah, I can't imagine such situation as well, but the behavior should be close to what we have in `AsyncResource`.

Reply:
Indeed. I was actually originally considering just using `AsyncResource` for here, but it currently only exists in the public file and doesn't allow for the split up before and after. (Though it probably wouldn't be too hard to adapt it to allow that.) I mostly just used a separate PromiseWrap class to have a fairly one-to-one port just to get it _working_ before doing any particular optimization.